### PR TITLE
fix(multiparser): respect 0-indent markdown-in-js

### LINKED ISSUE #3670: Prettier respects template literals, but not tagged templates
As far as I can tell, it's prettier's philosophy to leave template literals unaffected (except for breaking on the tokens) because adding breaks and indentation actually modifies the code. This tends to work fine, but it breaks when it comes to tagged templates, the syntax where you call a function with a template literal and leave off the parens.

A simple fix in my code was to add the parens back in to the function call, but this won't work for everyone as it actually changes the arguments for the resulting function call.

**Prettier 1.9.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBbAJgAwDpQCoAWAlgM4AEANkVHGSQRAK4XpQDkMZARrdegvHRkAhp2EUKZamRgFaAJzglmMagHMyTGAAdGMADTc9UzqQB0uHFFwYAFFYAKimKrjyy6CEplyy8omoEnLLqmlA+pJTUtMJQQtqKJG4Abt6yvHECokTQlHCpFBZQmACUuCD6IBDaqtAkyKDC8vIQAO4OTQj1KBUgwiQwyABm4kmVXPLCYADWcDAAytpT6sgw8oxwlZ5gw6ObINRJ8jBOwmqowrsUYyAAViQAHgBCkzNz88KocAAy0Vc3Wl0MAATP99kt5EdkCAuMIuABPCjQXoJagwADqRHQsmQAA4AAyVBIQJLoybaaEJJQpOC9RQAR0YREUp3OlyQI2u+ySqCIq3W3PUFDgAEVGBB4GDKjA4ZjsQRkMDpZMiFQoGoAMIQVAXaFQaC0yqMJJ4OHdTlJAC+lqAA)

**Input:**
```jsx
md`
This line shouldn't be indented at all in the resulting output, but it is.
`

md(`
Prettier does the right thing on this line and preserves the indentation level.
`)
```

**Output:**
```jsx
md`
  This line shouldn't be indented at all in the resulting output, but it is.
`;

md(`
Prettier does the right thing on this line and preserves the indentation level.
`)
```

### LINKED ISSUE #3740: Markdown: template string in codeblock is broken
**Prettier 1.10.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBaABARgDpXejAJl33wwGYTT8ADOgKwGcrrGIBbOGACwEsoA5jRakAhowAmuEfnFSoM9HMXC81dHVUgANCAgAHGL2iNkofQCd+MAOq8JPZAA4ADLpiiARnYfdkhXUZ+AQAbOABFAFcIeGQAM1EQxjh3C1FeEOCAYQ52UWQQKGg4HRBPNLAAay4AZX1RMGDkGAtIlJAmAA8AIQrqmBrRTgAZfhKkBKT2+otkiwK8i0qJCAB3KFLk9l5m1vbI5IAVL1MJxOTdfjmYAAU0gTz48-aLOABHSN5Xu9EH-LOprpLBBkjY0voCpY4HMAG4lXSiCwWNY3REIU4gUQwiD2UorMBPQGYxgwQkXPSRGD6SmEMlwAC+9KAA)
```sh
--parser markdown
```

**Input:**
````markdown
- 1
  - 2
    - 3
      ```js
      something`
      asd

      asd

      asd
      `
      ```

````

**Output:**
````markdown
* 1

  * 2

    * 3

      ```js
      something`
asd

asd

asd
`;
      ```

````

### PR DESCRIPTION
**Pull request by ikatyang**
Fixes #3670 

<details>
<summary>playground</summary>

**Prettier pr-3676**
[Playground link](https://deploy-preview-3676--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEBbAJgAwDpQCoAWAlgM4AEANkVHGSQRAK4XpQDkMZARrdegvHRkAhp2EUKZamRgFaAJzglmMagHMyTGAAdGMADTc9UzqQB0uHFFwYAFFYAKimKrjyy6CEplyy8omoEnLLqmlA+pJTUtMJQQtqKJG4Abt6yvHECokTQlHCpFBZQmACUIPogENqq0CTIoMLy8hAA7g6NCHUowskQROjlIMIkMMgAZuJJFVzywmAA1nAwAMrac+rIMPKMcBWeYOOTuyDUSfIwTsJqqMKHFFMgAFYkAB4AQrMLS8vCqHAAMtE7g8tLoYAAmYHHNbyM7IEBcYRcACeFGggwS1BgAHV+rJkAAOAAMFQSECS2Nm2nhCSUKTgg0UAEdGERFJdrrckBN7sckqgiJttnz1BQ4ABFRgQeBQiowJG49D4pDguWzIhUKBqADCEFQN3hUGgDIqjCSeCRXR5SQAvjagA)

**Input:**
```jsx
md`
This line shouldn't be indented at all in the resulting output, but it is.
`

md(`
Prettier does the right thing on this line and preserves the indentation level.
`)
```

**Output:**
```jsx
md`
This line shouldn't be indented at all in the resulting output, but it is.
`;

md(`
Prettier does the right thing on this line and preserves the indentation level.
`);

```

</details>

---

Fixes #3740

<details>
<summary>playground</summary>

**Prettier pr-3676**
[Playground link](https://deploy-preview-3676--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEBaABARgDpXejAJl33wwGYTT8ADOgKwGcrrGIBbOGACwEsoA5jRakAhowAmuEfnFSoM9HMXC81dHRogANCAgAHGL2iNkoUQCcLEAO4AFSwlMpRANwi8JOkOJjIAZqIANoxwugBGFqJgANZcAMr60fwCyDAWAK5hIBIQYAHBobr8oRYwdlEC7KIFIdlMAB4AQlGxCaKcADL8cLVFehkw+oOEfdlJFqXIINUWMbk2UN76FvwwAOqePMgAHAAMuisQoetR+tMrcKWuvboWcACOGbz3FaJVNUiBdbqh7LxpTLZRgpIJwACKGQg8DGuhgonCmwk2yQhDhUV4QRSAGEONVplBoLcQBlQgAVBHOb6hAC+NKAA)
```sh
--parser markdown
```

**Input:**
````markdown
- 1
  - 2
    - 3
      ```js
      something`
      asd

      asd

      asd
      `
      ```
````

**Output:**
````markdown
* 1

  * 2

    * 3

      ```js
      something`
      asd
      
      asd
      
      asd
      `;
      ```

````

</details>

### CODE REVIEW

File: src/language-js/embed.js [hunk 1]
```diff
@@ -143,19 +143,18 @@ function embed(path, print, textToDoc /*, options */) {
...
-          }
-        );
+        const text = parent.quasis[0].value.cooked;
+        const indentation = getIndentation(text);
+        const hasIndent = indentation !== "";
         return concat([
-          indent(
+          (hasIndent ? indent : identity)(
             concat([
               softline,
```
Comment (duailibe) [inline src/language-js/embed.js]: I think you need to change this to `literalline` if no indent, see [playground link](https://deploy-preview-3676--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEBLAZgAgBQwE4CucAlJsADpSaYC2AJgAaXwAeMlDA3JQL6XWVBUEABoQEAA4xU0AM7JQAQzx4IAdwAKyhPJSKAbhFR1RIRbJjJ0igDay4YgEZ5FYANZwYAZQmvUUAHNkfCIxOggwK1t7MX97PBgNFwCaRSi7BxAAK1kWACEXd08vRRo4ABl-OHSY8QIYCXqAJhrM3zx45BBHRUcATxtoUwk8fxgAdWMYAAtkAA4ABjERiHtxlwkukbh4-WqxPDgARwJUQ6TFFLSkawyxexpUYMJM2X8AmzgARQIIeFaxDBepM6DNkE1AS5UDZ3gBhCA0VJdKDQfYgAj2AAqvV0t3sPB4QA)

File: tests/multiparser_js_markdown/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -1,5 +1,36 @@
...
+\`;
+
+if (true) {
+  md\`
+text1
+
+  text2
+
+  text3
+  \`;
```
Comment (duailibe) [inline tests/multiparser_js_markdown/__snapshots__/jsfmt.spec.js.snap]: I was just gonna comment this too :)

I think we need to split and join with `literalline`s, like in https://github.com/prettier/prettier/blob/dc4e44968f430e54eaa72fb55752dc78fd8df755/src/language-js/printer-estree.js#L1917-L1918

Comment (ikatyang) [reply tests/multiparser_js_markdown/__snapshots__/jsfmt.spec.js.snap]: Do we have `literal-line` or `literal-softline`? The linebreaks in markdown is not static. 🤔

Comment (duailibe) [reply tests/multiparser_js_markdown/__snapshots__/jsfmt.spec.js.snap]: I've made a `zeroIndent` primitive to print zero-indented docs. Do you want me to open a separate PR or push here?

Comment (duailibe) [reply tests/multiparser_js_markdown/__snapshots__/jsfmt.spec.js.snap]: I'll open a PR refactoring the template literal printing

File: src/language-js/embed.js [hunk 3]
```diff
@@ -143,19 +145,20 @@ function embed(path, print, textToDoc /*, options */) {
...
+        const text = parent.quasis[0].value.cooked;
+        const indentation = getIndentation(text);
+        const hasIndent = indentation !== "";
         return concat([
-          indent(
+          (hasIndent ? indent : identity)(
             concat([
-              softline,
-              docUtils.stripTrailingHardline(escapeBackticks(doc))
+              hasIndent ? softline : literalline,
```
Comment (duailibe) [inline src/language-js/embed.js]: I think all of these ternaries would be better replaced as an `if (hasIndent)`

File: src/language-js/embed.js [hunk 4]
```diff
@@ -165,12 +168,24 @@ function embed(path, print, textToDoc /*, options */) {
...
+
+  function printMarkdown(text) {
+    const doc = textToDoc(text, { parser: "markdown", __inJsTemplate: true });
+    return docUtils.stripTrailingHardline(escapeBackticks(doc));
+  }
+}
+
+function literalify(doc) {
+  return align(-Infinity, doc);
+}
```
Comment (j-f1) [inline src/language-js/embed.js]: Could you make this a “primitive” (like how `conditionalGroup` is actually a `group`) in the doc file?

Comment (ikatyang) [reply src/language-js/embed.js]: Not sure how should we name it, `rootIndent`?

**Comment by ikatyang** [comment]
I'd like to also fix that edge case in this PR, still WIP.

File: src/language-js/embed.js (hunk 4 above)
Comment (j-f1) [reply src/language-js/embed.js]: Sure. Or maybe `dedentToRoot`?

**Comment by ikatyang** [comment]
Ready for review.

**Comment by chrislloyd** [comment]
This is great, thank you very much! Would it be possible to get a point release cut with this commit?

**Comment by christianvuerings** [comment]
@duailibe @azz Do you know when we might get a next release which includes this fix?

**Comment by btraut** [comment]
@duailibe @azz @ikatyang Checking up on this again. Is there any chance we could get a fresh point release published? Thanks!

**Comment by lydell** [comment]
@btraut 1.11 has started to form: #3931

**Pull request merged**